### PR TITLE
feat(upload-api) Added new upload api service

### PR DESCRIPTION
**Pull request by edge33**
### Ticket:

- [18782](https://dev.azure.com/OGCIO-Digital-Services/Digital%20Services%20Programme/_boards/board/t/Life%20Events/Stories?System.AreaPath=Life%20Events&workitem=18782)

### Description

Initial versione of file upload service

## Type

- [ ] **Dependency upgrade**
- [ ] **Bug fix**
- [x] **New feature**
- [ ] **Dev change**

### Checklist:

- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] This change might impact the developer experience of others and should be communicated

### Screenshots:

N/A

### CODE REVIEW

File: apps/file-upload-api/openapi-definition.yml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  version: 0.1.0
+components:
+  schemas: {}
+paths:
+  /health:
+    get:
+      responses:
+        "200":
+          description: Default Response
+  /files/upload:
```
Comment (relvao) [inline apps/file-upload-api/openapi-definition.yml:14]: why not just /files?

File: apps/file-upload-api/src/app.ts [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+export async function build(opts?: FastifyServerOptions) {
+  const app = fastify(opts).withTypeProvider<TypeBoxTypeProvider>();
+  initializeLoggingHooks(app);
+  initializeErrorHandler(app);
+
+  // limit 20 gb file
+  app.register(multipart, {
+    limits: {
+      fileSize: 20 * 1024 * 1024 * 1024,
```
Comment (relvao) [inline apps/file-upload-api/src/app.ts:30]: can we add the file limit to an env var?

File: apps/file-upload-api/src/app.ts [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+      ),
+    },
+  });
+
+  app.register(postgres, {
+    host: app.config.POSTGRES_HOST,
+    port: Number(app.config.POSTGRES_PORT),
+    user: app.config.POSTGRES_USER,
+    password: app.config.POSTGRES_PASSWORD,
+    database: app.config.POSTGRES_DB_NAME_SHARED,
```
Comment (relvao) [inline apps/file-upload-api/src/app.ts:70]: we need to create a db for this project. Shouldn't use the shared db. Happy for it to be a follow up story

File: apps/file-upload-api/src/config.ts [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+  type: "object",
+  required: ["CLAMAV_HOST", "S3_ENDPOINT", "S3_REGION", "S3_BUCKET_NAME"],
+  properties: {
+    CLAMAV_HOST: { type: "string" },
+    S3_ENDPOINT: { type: "string" },
+    S3_REGION: { type: "string" },
+    S3_ACCESS_KEY_ID: { type: "string" },
+    S3_SECRET_ACCESS_KEY: { type: "string" },
+    S3_BUCKET_NAME: { type: "string" },
+  },
```
Comment (relvao) [inline apps/file-upload-api/src/config.ts:11]: shouldn't the db config be a part of this as well?

File: apps/file-upload-api/src/routes/files/index.ts [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+      return reject(app.httpErrors.badRequest("Filename not provided"));
+    }
+
+    const s3Config = app.s3Client;
+    let scanComplete = false;
+    let outputFinished = false;
+    let fileInfected = false;
+    const antivirusPassthrough = app.avClient.passthrough();
+    antivirusPassthrough.on("error", (err) => {
+      reject(app.httpErrors.internalServerError(err.message));
```
Comment (Eomm) [inline apps/file-upload-api/src/routes/files/index.ts:35]: I think this error would be bubble up to the pipeline error too - this means that we would run `reject` twice.

In this case, you may handle it with an internal `EventEmitter` and the `once` -> so you centralize who is gonna to run `resolve/reject`

File: apps/file-upload-api/src/app.ts (hunk 2 above)
Comment (Eomm) [inline apps/file-upload-api/src/app.ts:30]: Did you set the https://fastify.dev/docs/latest/Reference/Server/#requesttimeout ?

In this case I think you need to customize it with `request.raw.setTimeout()` for the upload routes

File: apps/file-upload-api/src/routes/files/index.ts (hunk 5 above)
Comment (edge33) [reply apps/file-upload-api/src/routes/files/index.ts:35]: I like this approach but still, there are cases where `antivirusPassthrough` will receive multiple `error` events, if we catch an error with `once` subsequent errors will be thrown. It is also the case that even if antivirus fails the pipeline will push the stream data to the upload that will fail in consequence. I might accept having multiple error logs, what do you think?

File: apps/file-upload-api/src/routes/files/index.ts [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+        app.log.error(err);
+      });
+
+      reply.send(downloadPassthrough);
+
+      const thePromise = new Promise<void>((resolve, reject) => {
+        antivirusPassthrough.once("scan-complete", (result) => {
+          const { isInfected } = result;
+          if (isInfected) {
+            reject("File is infected");
```
Comment (Eomm) [inline apps/file-upload-api/src/routes/files/index.ts:301]:
```suggestion
            reject("File is infected");
            return;
```

File: apps/file-upload-api/src/routes/files/index.ts [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+      const monitorPassThrough = new PromiseTransform(thePromise);
+
+      pipeline(
+        stream,
+        antivirusPassthrough,
+        monitorPassThrough,
+        downloadPassthrough,
+        (err) => {
+          if (err) {
+            downloadPassthrough.destroy();
```
Comment (Eomm) [inline apps/file-upload-api/src/routes/files/index.ts:315]: I would expect it is not necessary

https://nodejs.org/api/stream.html#streampipelinestreams-callback

stream.pipeline() will call stream.destroy(err) on all streams except:

Readable streams which have emitted 'end' or 'close'.
Writable streams which have emitted 'finish' or 'close'.

File: apps/file-upload-api/.gitignore [hunk 8]
```diff
@@ -0,0 +1,3 @@
+node_modules
+.env
+.tap
```
Comment (SamSalvatico) [inline apps/file-upload-api/.gitignore:3]: could you have a `dist` folder?

File: apps/file-upload-api/src/app.ts [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+    limits: {
+      fileSize: app.config.MAX_FILE_SIZE as number,
+    },
+  });
+
+  app.register(fastifySwagger, {
+    openapi: {
+      info: {
+        title: "OGCIO File Upload API",
+        description: "API for OGCIO file upload servce",
```
Comment (SamSalvatico) [inline apps/file-upload-api/src/app.ts:51]:
```suggestion
        description: "API for OGCIO file upload service",
```

File: apps/file-upload-api/src/config.ts [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+    "MAX_FILE_SIZE",
+  ],
+  properties: {
+    CLAMAV_HOST: { type: "string" },
+    S3_ENDPOINT: { type: "string" },
+    S3_REGION: { type: "string" },
+    S3_ACCESS_KEY_ID: { type: "string" },
+    S3_SECRET_ACCESS_KEY: { type: "string" },
+    S3_BUCKET_NAME: { type: "string" },
+    MAX_FILE_SIZE: { type: "number" },
```
Comment (SamSalvatico) [inline apps/file-upload-api/src/config.ts:17]: you are missing the `LOGTO_` keys

File: apps/file-upload-api/src/index.ts [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+import { getLoggingConfiguration } from "logging-wrapper";
+
+const app = await build(getLoggingConfiguration());
+
+app.listen({ host: "0.0.0.0", port: 8006 }, (err, address) => {
+  if (err) {
+    console.error(err);
+    process.exit(1);
+  }
+  console.log(`app listening at ${address}`);
```
Comment (SamSalvatico) [inline apps/file-upload-api/src/index.ts:13]:
```suggestion
  app.log.info(`app listening at ${address}`);
```

File: apps/file-upload-api/src/routes/files/index.ts [hunk 12]
```diff
@@ -0,0 +1,362 @@
...
+      Key: filename,
+    }),
+  );
+};
+
+const scanAndUplad = async (app: FastifyInstance, request: FastifyRequest) => {
+  const data = await request.file();
+
+  if (!data) {
+    throw app.httpErrors.badRequest("Request is not multipart");
```
Comment (SamSalvatico) [inline apps/file-upload-api/src/routes/files/index.ts:46]: please use the shared errors
`new BadRequest( ....`

File: apps/file-upload-api/src/routes/files/index.ts [hunk 13]
```diff
@@ -0,0 +1,362 @@
...
+  }
+
+  const stream = data.file;
+  const filename = data.filename;
+
+  if (!filename) {
+    throw app.httpErrors.badRequest("Filename not provided");
```
Comment (SamSalvatico) [inline apps/file-upload-api/src/routes/files/index.ts:53]: same

File: apps/file-upload-api/src/routes/files/index.ts [hunk 14]
```diff
@@ -43,35 +53,36 @@ const scanAndUplad = async (app: FastifyInstance, request: FastifyRequest) => {
...
+      reject(new CustomError(FILE_UPLOAD, "File is infected", 400));
     });
 
     eventEmitter.once("fileUploaded", () => {
       resolve();
     });
 
     eventEmitter.once("fileTooLarge", () => {
-      reject(app.httpErrors.badRequest("File too large"));
+      reject(new CustomError(FILE_UPLOAD, "File is too large", 400));
```
Comment (SamSalvatico) [inline apps/file-upload-api/src/routes/files/index.ts:78]: why you use Custom Error instead of BadRequest?

Comment (edge33) [reply apps/file-upload-api/src/routes/files/index.ts:78]: good catch!

**Review by SamSalvatico — APPROVED** [review]

**Pull request merged**